pkg/image/trigger/annotations/annotations.go

```diff
@@ -0,0 +1,273 @@
+package annotations
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+
+	ometa "github.com/openshift/origin/pkg/api/meta"
+	triggerapi "github.com/openshift/origin/pkg/image/api/v1/trigger"
+	"github.com/openshift/origin/pkg/image/trigger"
+)
+
+func calculateAnnotationTriggers(m metav1.Object, prefix string) (string, string, []triggerapi.ObjectFieldTrigger, error) {
+	var key, namespace string
+	if namespace = m.GetNamespace(); len(namespace) > 0 {
+		key = prefix + namespace + "/" + m.GetName()
+	} else {
+		key = prefix + m.GetName()
+	}
+	t, ok := m.GetAnnotations()[triggerapi.TriggerAnnotationKey]
+	if !ok {
+		return key, namespace, nil, nil
+	}
+	triggers := []triggerapi.ObjectFieldTrigger{}
+	if err := json.Unmarshal([]byte(t), &triggers); err != nil {
+		return key, namespace, nil, err
+	}
+	if hasDuplicateTriggers(triggers) {
+		return key, namespace, nil, fmt.Errorf("duplicate triggers are not allowed")
```

Comment:
The other discussion in the thread stands - I want to set a broader policy for it first, then come back and be deliberate.